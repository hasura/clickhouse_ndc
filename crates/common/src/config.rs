use std::collections::BTreeMap;

use crate::{
    clickhouse_parser::{datatype::ClickHouseDataType, parameterized_query::ParameterizedQuery},
    config_file::{ParameterizedQueryExposedAs, PrimaryKey},
};

#[derive(Debug, Clone)]
/// In memory, runtime configuration, built from the configuration file(s) and environment variables
pub struct ServerConfig {
    /// the connection part of the config is not part of the config file
    pub connection: ConnectionConfig,
    pub namespace_separator: String,
    pub table_types: BTreeMap<ReturnTypeRef, TableType>,
    pub tables: BTreeMap<String, TableConfig>,
    pub queries: BTreeMap<String, ParameterizedQueryConfig>,
}

#[derive(Debug, Clone)]
pub struct TableType {
    pub comment: Option<String>,
    pub columns: BTreeMap<String, ClickHouseDataType>,
}

#[derive(Debug, Default, Clone)]
pub struct ConnectionConfig {
    pub username: String,
    pub password: String,
    pub url: String,
}

#[derive(Debug, Clone)]
pub struct TableConfig {
    /// The table name
    pub name: String,
    /// The table schema
    pub schema: String,
    /// Comments are sourced from the database table comment
    pub comment: Option<String>,
    pub primary_key: Option<PrimaryKey>,
    pub arguments: BTreeMap<String, ClickHouseDataType>,
    // this key coresponds to a return type definition in the config table types
    pub return_type: ReturnTypeRef,
}

#[derive(Debug, Clone)]
pub struct ParameterizedQueryConfig {
    pub exposed_as: ParameterizedQueryExposedAs,
    pub comment: Option<String>,
    pub query: ParameterizedQuery,
    pub return_type: ReturnTypeRef,
}

type ReturnTypeRef = String;
