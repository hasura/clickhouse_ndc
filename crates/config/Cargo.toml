[package]
name = "config"
version.workspace = true
edition.workspace = true

[dependencies]
schemars = "0.8.16"
serde = { version = "1.0.197", features = ["derive"] }
