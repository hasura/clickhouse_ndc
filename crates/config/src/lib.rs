use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

#[derive(Debug, Default, Clone, Serialize, Deserialize, JsonSchema)]
pub struct ServerConfigFile {
    #[serde(rename = "$schema")]
    pub schema: String,
    /// A list of tables available in this database
    pub tables: Vec<TableConfig>,
}

#[derive(Debug, Default, Clone, Serialize, Deserialize, JsonSchema)]
pub struct ServerConfig {
    /// the connection part of the config is not part of the config file
    pub connection: ConnectionConfig,
    pub tables: Vec<TableConfig>,
}

#[derive(Debug, Default, Clone, Serialize, Deserialize, JsonSchema)]
pub struct ConnectionConfig {
    pub username: String,
    pub password: String,
    pub url: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct TableConfig {
    /// The table name
    pub name: String,
    /// The table schema
    pub schema: String,
    /// The table alias defaults to "<table_schema>_<table_name>", except for tables in the "default" schema where the table name is used
    /// This is the name exposed to the engine, and may be configured by users. This is preserved through config updates
    pub alias: String,
    /// Comments are sourced from the database table comment
    pub comment: Option<String>,
    pub primary_key: Option<PrimaryKey>,
    pub columns: Vec<ColumnConfig>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct PrimaryKey {
    pub name: String,
    pub columns: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct ColumnConfig {
    /// The column name
    pub name: String,
    /// The column alias defaults to the column name, but may be changed by users. This is preserved through config updates
    pub alias: String,
    /// The column data type
    pub data_type: String,
}

pub const CONFIG_FILE_NAME: &str = "configuration.json";
pub const CONFIG_SCHEMA_FILE_NAME: &str = "configuration.schema.json";
