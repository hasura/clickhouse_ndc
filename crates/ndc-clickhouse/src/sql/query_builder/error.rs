use std::fmt;

use ndc_sdk::connector::{ExplainError, QueryError};

use super::typecasting::TypeStringError;

#[derive(Debug, PartialEq)]
pub enum QueryBuilderError {
    /// A relationship referenced in the query is missing from the collection_relationships map
    MissingRelationship(String),
    /// An argument required for a native query was not supplied
    MissingNativeQueryArgument { query: String, argument: String },
    /// A table was referenced but not found in configuration
    UnknownTable(String),
    /// An argument was supplied for a table that does not have that argument
    UnknownTableArgument { table: String, argument: String },
    /// An argument was supplied for a table that does not have that argument
    UnknownQueryArgument { query: String, argument: String },
    /// A table in configuration referenced a table type that could not be found
    UnknownTableType(String),
    /// A column was referenced but not found in configuration
    UnknownColumn(String, String),
    /// A field was referenced but not found in configuration
    UnknownSubField {
        field_name: String,
        data_type: String,
    },
    /// Unable to serialize variables into a json string
    CannotSerializeVariables(String),
    /// An unknown single column aggregate function was referenced
    UnknownSingleColumnAggregateFunction(String),
    /// An unknown binary comparison operator was referenced
    UnknownBinaryComparisonOperator(String),
    /// A feature is not supported
    NotSupported(String),
    /// An error that should never happen, and indicates a bug if triggered
    Unexpected(String),
    /// There was an issue creating typecasting strings
    Typecasting(TypeStringError),
    /// Column type did not match type asserted by request
    ColumnTypeMismatch { expected: String, got: String },
}

impl fmt::Display for QueryBuilderError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            QueryBuilderError::MissingRelationship(rel) => write!(f, "Missing relationship: {rel}"),
            QueryBuilderError::MissingNativeQueryArgument { query, argument } => write!(
                f,
                "Argument {argument} required for native query {query} was not supplied"
            ),
            QueryBuilderError::UnknownTable(t) => write!(f, "Unable to find table {t} in config"),
            QueryBuilderError::UnknownTableArgument { table, argument } => {
                write!(f, "Unknown argument {argument} supplied for table {table}")
            }
            QueryBuilderError::UnknownQueryArgument { query, argument } => {
                write!(f, "Unknown argument {argument} supplied for query {query}")
            }

            QueryBuilderError::UnknownTableType(t) => {
                write!(f, "Unable to find table type {t} in config")
            }
            QueryBuilderError::UnknownColumn(c, t) => {
                write!(f, "Unable to find column {c} for table {t} in config")
            }
            QueryBuilderError::UnknownSubField {
                field_name,
                data_type,
            } => {
                write!(f, "Unknown field {field_name} in type {data_type}")
            }
            QueryBuilderError::CannotSerializeVariables(e) => {
                write!(f, "Unable to serialize variables into a json string: {e}")
            }
            QueryBuilderError::UnknownSingleColumnAggregateFunction(agg) => {
                write!(f, "Unknown single column aggregate function: {agg}")
            }
            QueryBuilderError::UnknownBinaryComparisonOperator(op) => {
                write!(f, "Unknown binary comparison operator: {op}")
            }
            QueryBuilderError::NotSupported(e) => write!(f, "Not supported: {e}"),
            QueryBuilderError::Unexpected(e) => write!(f, "Unexpected: {e}"),
            QueryBuilderError::Typecasting(e) => write!(f, "Typecasting: {e}"),
            QueryBuilderError::ColumnTypeMismatch { expected, got } => {
                write!(f, "Column Type Mismatch: expected {expected}, got {got}")
            }
        }
    }
}

impl std::error::Error for QueryBuilderError {}

impl From<QueryBuilderError> for QueryError {
    fn from(value: QueryBuilderError) -> Self {
        match value {
            QueryBuilderError::MissingRelationship(_)
            | QueryBuilderError::MissingNativeQueryArgument { .. }
            | QueryBuilderError::UnknownTable(_)
            | QueryBuilderError::UnknownTableArgument { .. }
            | QueryBuilderError::UnknownQueryArgument { .. }
            | QueryBuilderError::UnknownTableType(_)
            | QueryBuilderError::UnknownColumn(_, _)
            | QueryBuilderError::UnknownSubField { .. }
            | QueryBuilderError::CannotSerializeVariables(_)
            | QueryBuilderError::UnknownSingleColumnAggregateFunction(_)
            | QueryBuilderError::UnknownBinaryComparisonOperator(_)
            | QueryBuilderError::Typecasting(_)
            | QueryBuilderError::ColumnTypeMismatch { .. } => {
                QueryError::new_invalid_request(&value)
            }
            QueryBuilderError::NotSupported(_) => QueryError::new_unsupported_operation(&value),
            QueryBuilderError::Unexpected(_) => QueryError::new(value),
        }
    }
}

impl From<QueryBuilderError> for ExplainError {
    fn from(value: QueryBuilderError) -> Self {
        match value {
            QueryBuilderError::MissingRelationship(_)
            | QueryBuilderError::MissingNativeQueryArgument { .. }
            | QueryBuilderError::UnknownTable(_)
            | QueryBuilderError::UnknownTableArgument { .. }
            | QueryBuilderError::UnknownQueryArgument { .. }
            | QueryBuilderError::UnknownTableType(_)
            | QueryBuilderError::UnknownColumn(_, _)
            | QueryBuilderError::UnknownSubField { .. }
            | QueryBuilderError::CannotSerializeVariables(_)
            | QueryBuilderError::UnknownSingleColumnAggregateFunction(_)
            | QueryBuilderError::UnknownBinaryComparisonOperator(_)
            | QueryBuilderError::Typecasting(_)
            | QueryBuilderError::ColumnTypeMismatch { .. } => {
                ExplainError::new_invalid_request(&value)
            }
            QueryBuilderError::NotSupported(_) => ExplainError::new_unsupported_operation(&value),
            QueryBuilderError::Unexpected(_) => ExplainError::new(Box::new(value)),
        }
    }
}
