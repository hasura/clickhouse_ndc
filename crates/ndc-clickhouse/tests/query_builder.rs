use common::config_file::ServerConfigFile;
use ndc_clickhouse::sql::QueryBuilderError;
use ndc_sdk::models;
use schemars::schema_for;
use std::error::Error;
use tokio::fs;

mod test_utils {
    use common::config::ServerConfig;
    use ndc_clickhouse::{
        connector::{handler::schema_response, setup::ClickhouseConnectorSetup},
        sql::{QueryBuilder, QueryBuilderError},
    };
    use ndc_sdk::models::{self, SchemaResponse};
    use std::{collections::HashMap, env, error::Error, path::PathBuf};
    use tokio::fs;

    /// when running tests locally, this can be set to true to update reference files
    /// this allows us to view diffs between commited samples and fresh samples
    /// we don't want that behavior when running CI, so this value should be false in commited code
    const UPDATE_SNAPSHOTS: bool = false;

    fn base_path(schema_dir: &str) -> PathBuf {
        PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .join("tests")
            .join("query_builder")
            .join(schema_dir)
    }
    fn tests_dir_path(schema_dir: &str, group_dir: &str) -> PathBuf {
        base_path(schema_dir).join(group_dir)
    }
    fn config_dir_path(schema_dir: &str) -> PathBuf {
        base_path(schema_dir).join("_config")
    }
    fn expected_schema_path(schema_dir: &str) -> PathBuf {
        base_path(schema_dir).join("_schema").join("schema.json")
    }
    async fn read_mock_configuration(schema_dir: &str) -> Result<ServerConfig, Box<dyn Error>> {
        // set mock values for required env vars, we won't be reading these anyways
        let env = HashMap::from_iter(vec![
            ("CLICKHOUSE_URL".to_owned(), "".to_owned()),
            ("CLICKHOUSE_USERNAME".to_owned(), "".to_owned()),
            ("CLICKHOUSE_PASSWORD".to_owned(), "".to_owned()),
        ]);
        let setup = ClickhouseConnectorSetup::new_from_env(env);
        let config_dir = config_dir_path(schema_dir);
        let configuration = setup.read_server_config(config_dir).await?;
        Ok(configuration)
    }
    async fn read_request(
        schema_dir: &str,
        group_dir: &str,
        test_name: &str,
    ) -> Result<models::QueryRequest, Box<dyn Error>> {
        let request_path =
            tests_dir_path(schema_dir, group_dir).join(format!("{test_name}.request.json"));

        let file_content = fs::read_to_string(request_path).await?;
        let request: models::QueryRequest = serde_json::from_str(&file_content)?;

        Ok(request)
    }
    async fn read_expected_sql(
        schema_dir: &str,
        group_dir: &str,
        test_name: &str,
    ) -> Result<String, Box<dyn Error>> {
        let statement_path =
            tests_dir_path(schema_dir, group_dir).join(format!("{test_name}.statement.sql"));
        let expected_statement = fs::read_to_string(&statement_path).await?;
        Ok(expected_statement)
    }
    async fn write_expected_sql(
        schema_dir: &str,
        group_dir: &str,
        test_name: &str,
        generated_statement: &str,
    ) -> Result<(), Box<dyn Error>> {
        let statement_path =
            tests_dir_path(schema_dir, group_dir).join(format!("{test_name}.statement.sql"));
        let pretty_statement = pretty_print_sql(generated_statement);
        fs::write(&statement_path, &pretty_statement).await?;
        Ok(())
    }
    fn pretty_print_sql(query: &str) -> String {
        use sqlformat::{format, FormatOptions, Indent, QueryParams};
        let params = QueryParams::None;
        let options = FormatOptions {
            indent: Indent::Spaces(2),
            uppercase: false,
            lines_between_queries: 1,
        };

        format(query, &params, options)
    }
    fn generate_sql(
        configuration: &ServerConfig,
        request: &models::QueryRequest,
    ) -> Result<String, QueryBuilderError> {
        let generated_statement = pretty_print_sql(
            &QueryBuilder::new(request, configuration)
                .build_inlined()?
                .to_string(),
        );
        Ok(generated_statement)
    }
    pub async fn test_generated_sql(
        schema_dir: &str,
        group_dir: &str,
        test_name: &str,
    ) -> Result<(), Box<dyn Error>> {
        let configuration = read_mock_configuration(schema_dir).await?;
        let request = read_request(schema_dir, group_dir, test_name).await?;

        let generated_sql = generate_sql(&configuration, &request)?;

        if UPDATE_SNAPSHOTS {
            write_expected_sql(schema_dir, group_dir, test_name, &generated_sql).await?;
        } else {
            let expected_sql = read_expected_sql(schema_dir, group_dir, test_name).await?;

            assert_eq!(generated_sql, expected_sql);
        }

        Ok(())
    }
    pub async fn test_error(
        schema_dir: &str,
        group_dir: &str,
        test_name: &str,
        err: QueryBuilderError,
    ) -> Result<(), Box<dyn Error>> {
        let configuration = read_mock_configuration(schema_dir).await?;
        let request = read_request(schema_dir, group_dir, test_name).await?;

        let result = generate_sql(&configuration, &request);

        assert_eq!(result, Err(err));

        Ok(())
    }
    pub async fn test_schema(schema_dir: &str) -> Result<(), Box<dyn Error>> {
        let configuration = read_mock_configuration(schema_dir).await?;
        let schema = schema_response(&configuration);
        let expected_schema_path = expected_schema_path(schema_dir);

        if UPDATE_SNAPSHOTS {
            fs::write(expected_schema_path, serde_json::to_string_pretty(&schema)?).await?;
        } else {
            let expected_schema_file = fs::read_to_string(expected_schema_path).await?;
            let expected_schema: SchemaResponse = serde_json::from_str(&expected_schema_file)?;

            assert_eq!(schema, expected_schema, "Schema should match snapshot")
        }

        Ok(())
    }
}

#[tokio::test]
#[ignore]
async fn update_json_schema() -> Result<(), Box<dyn Error>> {
    fs::write(
        "./tests/query_builder/request.schema.json",
        serde_json::to_string_pretty(&schema_for!(models::QueryRequest))?,
    )
    .await?;
    fs::write(
        "./tests/query_builder/configuration.schema.json",
        serde_json::to_string_pretty(&schema_for!(ServerConfigFile))?,
    )
    .await?;

    Ok(())
}

#[cfg(test)]
mod schemas {
    use super::*;

    #[tokio::test]
    async fn chinook_schema() -> Result<(), Box<dyn Error>> {
        test_utils::test_schema("chinook").await
    }
    #[tokio::test]
    async fn complex_columns_schema() -> Result<(), Box<dyn Error>> {
        test_utils::test_schema("complex_columns").await
    }
    #[tokio::test]
    async fn star_schema_schema() -> Result<(), Box<dyn Error>> {
        test_utils::test_schema("star_schema").await
    }
}

#[cfg(test)]
mod simple_queries {
    use super::*;

    async fn test_generated_sql(name: &str) -> Result<(), Box<dyn Error>> {
        super::test_utils::test_generated_sql("chinook", "01_simple_queries", name).await
    }

    #[tokio::test]
    async fn select_rows() -> Result<(), Box<dyn Error>> {
        test_generated_sql("01_select_rows").await
    }
    #[tokio::test]
    async fn with_predicate() -> Result<(), Box<dyn Error>> {
        test_generated_sql("02_with_predicate").await
    }
    #[tokio::test]
    async fn larger_predicate() -> Result<(), Box<dyn Error>> {
        test_generated_sql("03_larger_predicate").await
    }
    #[tokio::test]
    async fn limit() -> Result<(), Box<dyn Error>> {
        test_generated_sql("04_limit").await
    }
    #[tokio::test]
    async fn offset() -> Result<(), Box<dyn Error>> {
        test_generated_sql("05_offset").await
    }
    #[tokio::test]
    async fn limit_offset() -> Result<(), Box<dyn Error>> {
        test_generated_sql("06_limit_offset").await
    }

    #[tokio::test]
    async fn order_by() -> Result<(), Box<dyn Error>> {
        test_generated_sql("07_order_by").await
    }
    #[tokio::test]
    async fn predicate_limit_offset_order_by() -> Result<(), Box<dyn Error>> {
        test_generated_sql("08_predicate_limit_offset_order_by").await
    }
}

#[cfg(test)]
mod relationships {
    use super::*;

    async fn test_generated_sql(name: &str) -> Result<(), Box<dyn Error>> {
        super::test_utils::test_generated_sql("chinook", "02_relationships", name).await
    }

    #[tokio::test]
    async fn object_relationship() -> Result<(), Box<dyn Error>> {
        test_generated_sql("01_object_relationship").await
    }
    #[tokio::test]
    async fn array_relationship() -> Result<(), Box<dyn Error>> {
        test_generated_sql("02_array_relationship").await
    }
    #[tokio::test]
    async fn parent_predicate() -> Result<(), Box<dyn Error>> {
        test_generated_sql("03_parent_predicate").await
    }
    #[tokio::test]
    async fn child_predicate() -> Result<(), Box<dyn Error>> {
        test_generated_sql("04_child_predicate").await
    }
    #[tokio::test]
    async fn traverse_relationship_in_predicate() -> Result<(), Box<dyn Error>> {
        test_generated_sql("05_traverse_relationship_in_predicate").await
    }
    #[tokio::test]
    async fn traverse_relationship_in_order_by() -> Result<(), Box<dyn Error>> {
        test_generated_sql("06_traverse_relationship_in_order_by").await
    }
    #[tokio::test]
    async fn order_by_aggregate_across_relationships() -> Result<(), Box<dyn Error>> {
        test_generated_sql("07_order_by_aggregate_across_relationships").await
    }
}

#[cfg(test)]
mod variables {
    use super::*;

    async fn test_generated_sql(name: &str) -> Result<(), Box<dyn Error>> {
        super::test_utils::test_generated_sql("chinook", "03_variables", name).await
    }

    #[tokio::test]
    async fn simple_predicate() -> Result<(), Box<dyn Error>> {
        test_generated_sql("01_simple_predicate").await
    }
    #[tokio::test]
    async fn empty_variable_sets() -> Result<(), Box<dyn Error>> {
        test_generated_sql("02_empty_variable_sets").await
    }
    #[tokio::test]
    async fn singe_set() -> Result<(), Box<dyn Error>> {
        test_generated_sql("03_single_set").await
    }
}

#[cfg(test)]
mod native_queries {
    use super::*;

    async fn test_generated_sql(name: &str) -> Result<(), Box<dyn Error>> {
        super::test_utils::test_generated_sql("star_schema", "01_native_queries", name).await
    }

    #[tokio::test]
    async fn native_query() -> Result<(), Box<dyn Error>> {
        test_generated_sql("01_native_query").await
    }
}

#[cfg(test)]
mod field_selector {
    use super::*;

    async fn test_generated_sql(name: &str) -> Result<(), Box<dyn Error>> {
        super::test_utils::test_generated_sql("complex_columns", "field_selector", name).await
    }
    async fn test_error(name: &str, err: QueryBuilderError) -> Result<(), Box<dyn Error>> {
        super::test_utils::test_error("complex_columns", "field_selector", name, err).await
    }

    #[tokio::test]
    async fn generate_column_accessor() -> Result<(), Box<dyn Error>> {
        test_generated_sql("01_generate_column_accessor").await
    }
    #[tokio::test]
    async fn skip_if_not_required() -> Result<(), Box<dyn Error>> {
        test_generated_sql("02_skip_if_not_required").await
    }
    #[tokio::test]
    async fn support_relationships_on_nested_field() -> Result<(), Box<dyn Error>> {
        test_generated_sql("03_support_relationships_on_nested_field").await
    }
    /// We do not support relationships on nested fileds if an array has been traversed
    #[tokio::test]
    async fn error_on_relationships_on_array_nested_field() -> Result<(), Box<dyn Error>> {
        let err = QueryBuilderError::NotSupported(
            "Relationships with fields nested in arrays".to_string(),
        );
        test_error("04_error_on_relationships_on_array_nested_field", err).await
    }
    #[tokio::test]
    async fn complex_example() -> Result<(), Box<dyn Error>> {
        test_generated_sql("05_complex_example").await
    }
    #[tokio::test]
    async fn no_useless_nested_accessors() -> Result<(), Box<dyn Error>> {
        test_generated_sql("06_no_useless_nested_accessors").await
    }
}
