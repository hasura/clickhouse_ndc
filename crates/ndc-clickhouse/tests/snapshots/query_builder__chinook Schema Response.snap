---
source: crates/ndc-clickhouse/tests/query_builder.rs
expression: schema
---
scalar_types:
  Date32:
    representation:
      type: string
    aggregate_functions:
      max:
        result_type:
          type: named
          name: Date32
      min:
        result_type:
          type: named
          name: Date32
    comparison_operators:
      _eq:
        type: equal
      _gt:
        type: custom
        argument_type:
          type: named
          name: Date32
      _gte:
        type: custom
        argument_type:
          type: named
          name: Date32
      _in:
        type: in
      _lt:
        type: custom
        argument_type:
          type: named
          name: Date32
      _lte:
        type: custom
        argument_type:
          type: named
          name: Date32
      _neq:
        type: custom
        argument_type:
          type: named
          name: Date32
      _nin:
        type: custom
        argument_type:
          type: array
          element_type:
            type: named
            name: Date32
  DateTime64(9):
    representation:
      type: string
    aggregate_functions:
      max:
        result_type:
          type: named
          name: DateTime64(9)
      min:
        result_type:
          type: named
          name: DateTime64(9)
    comparison_operators:
      _eq:
        type: equal
      _gt:
        type: custom
        argument_type:
          type: named
          name: DateTime64(9)
      _gte:
        type: custom
        argument_type:
          type: named
          name: DateTime64(9)
      _in:
        type: in
      _lt:
        type: custom
        argument_type:
          type: named
          name: DateTime64(9)
      _lte:
        type: custom
        argument_type:
          type: named
          name: DateTime64(9)
      _neq:
        type: custom
        argument_type:
          type: named
          name: DateTime64(9)
      _nin:
        type: custom
        argument_type:
          type: array
          element_type:
            type: named
            name: DateTime64(9)
  Float64:
    representation:
      type: float64
    aggregate_functions:
      avg:
        result_type:
          type: named
          name: Float64
      max:
        result_type:
          type: named
          name: Float64
      min:
        result_type:
          type: named
          name: Float64
      stddev_pop:
        result_type:
          type: named
          name: Float64
      stddev_samp:
        result_type:
          type: named
          name: Float64
      sum:
        result_type:
          type: named
          name: Float64
      var_pop:
        result_type:
          type: named
          name: Float64
      var_samp:
        result_type:
          type: named
          name: Float64
    comparison_operators:
      _eq:
        type: equal
      _gt:
        type: custom
        argument_type:
          type: named
          name: Float64
      _gte:
        type: custom
        argument_type:
          type: named
          name: Float64
      _in:
        type: in
      _lt:
        type: custom
        argument_type:
          type: named
          name: Float64
      _lte:
        type: custom
        argument_type:
          type: named
          name: Float64
      _neq:
        type: custom
        argument_type:
          type: named
          name: Float64
      _nin:
        type: custom
        argument_type:
          type: array
          element_type:
            type: named
            name: Float64
  Int32:
    representation:
      type: int32
    aggregate_functions:
      avg:
        result_type:
          type: named
          name: Float64
      max:
        result_type:
          type: named
          name: Int32
      min:
        result_type:
          type: named
          name: Int32
      stddev_pop:
        result_type:
          type: named
          name: Float64
      stddev_samp:
        result_type:
          type: named
          name: Float64
      sum:
        result_type:
          type: named
          name: Int64
      var_pop:
        result_type:
          type: named
          name: Float64
      var_samp:
        result_type:
          type: named
          name: Float64
    comparison_operators:
      _eq:
        type: equal
      _gt:
        type: custom
        argument_type:
          type: named
          name: Int32
      _gte:
        type: custom
        argument_type:
          type: named
          name: Int32
      _in:
        type: in
      _lt:
        type: custom
        argument_type:
          type: named
          name: Int32
      _lte:
        type: custom
        argument_type:
          type: named
          name: Int32
      _neq:
        type: custom
        argument_type:
          type: named
          name: Int32
      _nin:
        type: custom
        argument_type:
          type: array
          element_type:
            type: named
            name: Int32
  String:
    representation:
      type: string
    aggregate_functions: {}
    comparison_operators:
      _eq:
        type: equal
      _gt:
        type: custom
        argument_type:
          type: named
          name: String
      _gte:
        type: custom
        argument_type:
          type: named
          name: String
      _ilike:
        type: custom
        argument_type:
          type: named
          name: String
      _in:
        type: in
      _like:
        type: custom
        argument_type:
          type: named
          name: String
      _lt:
        type: custom
        argument_type:
          type: named
          name: String
      _lte:
        type: custom
        argument_type:
          type: named
          name: String
      _match:
        type: custom
        argument_type:
          type: named
          name: String
      _neq:
        type: custom
        argument_type:
          type: named
          name: String
      _nilike:
        type: custom
        argument_type:
          type: named
          name: String
      _nin:
        type: custom
        argument_type:
          type: array
          element_type:
            type: named
            name: String
      _nlike:
        type: custom
        argument_type:
          type: named
          name: String
object_types:
  Chinook_Album:
    description: ""
    fields:
      AlbumId:
        type:
          type: named
          name: Int32
      ArtistId:
        type:
          type: named
          name: Int32
      Title:
        type:
          type: named
          name: String
  Chinook_Artist:
    description: ""
    fields:
      ArtistId:
        type:
          type: named
          name: Int32
      Name:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
  Chinook_Customer:
    description: ""
    fields:
      Address:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
      City:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
      Company:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
      Country:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
      CustomerId:
        type:
          type: named
          name: Int32
      Email:
        type:
          type: named
          name: String
      Fax:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
      FirstName:
        type:
          type: named
          name: String
      LastName:
        type:
          type: named
          name: String
      Phone:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
      PostalCode:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
      State:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
      SupportRepId:
        type:
          type: nullable
          underlying_type:
            type: named
            name: Int32
  Chinook_Employee:
    description: ""
    fields:
      Address:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
      BirthDate:
        type:
          type: nullable
          underlying_type:
            type: named
            name: Date32
      City:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
      Country:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
      Email:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
      EmployeeId:
        type:
          type: named
          name: Int32
      Fax:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
      FirstName:
        type:
          type: named
          name: String
      HireDate:
        type:
          type: nullable
          underlying_type:
            type: named
            name: Date32
      LastName:
        type:
          type: named
          name: String
      Phone:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
      PostalCode:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
      ReportsTo:
        type:
          type: nullable
          underlying_type:
            type: named
            name: Int32
      State:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
      Title:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
  Chinook_Genre:
    description: ""
    fields:
      GenreId:
        type:
          type: named
          name: Int32
      Name:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
  Chinook_Invoice:
    description: ""
    fields:
      BillingAddress:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
      BillingCity:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
      BillingCountry:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
      BillingPostalCode:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
      BillingState:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
      CustomerId:
        type:
          type: named
          name: Int32
      InvoiceDate:
        type:
          type: named
          name: DateTime64(9)
      InvoiceId:
        type:
          type: named
          name: Int32
      Total:
        type:
          type: named
          name: Float64
  Chinook_InvoiceLine:
    description: ""
    fields:
      InvoiceId:
        type:
          type: named
          name: Int32
      InvoiceLineId:
        type:
          type: named
          name: Int32
      Quantity:
        type:
          type: named
          name: Int32
      TrackId:
        type:
          type: named
          name: Int32
      UnitPrice:
        type:
          type: named
          name: Float64
  Chinook_MediaType:
    description: ""
    fields:
      MediaTypeId:
        type:
          type: named
          name: Int32
      Name:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
  Chinook_Playlist:
    description: ""
    fields:
      Name:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
      PlaylistId:
        type:
          type: named
          name: Int32
  Chinook_PlaylistTrack:
    description: ""
    fields:
      PlaylistId:
        type:
          type: named
          name: Int32
      TrackId:
        type:
          type: named
          name: Int32
  Chinook_Track:
    description: ""
    fields:
      AlbumId:
        type:
          type: nullable
          underlying_type:
            type: named
            name: Int32
      Bytes:
        type:
          type: nullable
          underlying_type:
            type: named
            name: Int32
      Composer:
        type:
          type: nullable
          underlying_type:
            type: named
            name: String
      GenreId:
        type:
          type: nullable
          underlying_type:
            type: named
            name: Int32
      MediaTypeId:
        type:
          type: named
          name: Int32
      Milliseconds:
        type:
          type: named
          name: Int32
      Name:
        type:
          type: named
          name: String
      TrackId:
        type:
          type: named
          name: Int32
      UnitPrice:
        type:
          type: named
          name: Float64
collections:
  - name: Chinook_Album
    description: ""
    arguments: {}
    type: Chinook_Album
    uniqueness_constraints:
      AlbumId:
        unique_columns:
          - AlbumId
    foreign_keys: {}
  - name: Chinook_Artist
    description: ""
    arguments: {}
    type: Chinook_Artist
    uniqueness_constraints:
      ArtistId:
        unique_columns:
          - ArtistId
    foreign_keys: {}
  - name: Chinook_Customer
    description: ""
    arguments: {}
    type: Chinook_Customer
    uniqueness_constraints:
      CustomerId:
        unique_columns:
          - CustomerId
    foreign_keys: {}
  - name: Chinook_Employee
    description: ""
    arguments: {}
    type: Chinook_Employee
    uniqueness_constraints:
      EmployeeId:
        unique_columns:
          - EmployeeId
    foreign_keys: {}
  - name: Chinook_Genre
    description: ""
    arguments: {}
    type: Chinook_Genre
    uniqueness_constraints:
      GenreId:
        unique_columns:
          - GenreId
    foreign_keys: {}
  - name: Chinook_Invoice
    description: ""
    arguments: {}
    type: Chinook_Invoice
    uniqueness_constraints:
      InvoiceId:
        unique_columns:
          - InvoiceId
    foreign_keys: {}
  - name: Chinook_InvoiceLine
    description: ""
    arguments: {}
    type: Chinook_InvoiceLine
    uniqueness_constraints:
      InvoiceLineId:
        unique_columns:
          - InvoiceLineId
    foreign_keys: {}
  - name: Chinook_MediaType
    description: ""
    arguments: {}
    type: Chinook_MediaType
    uniqueness_constraints:
      MediaTypeId:
        unique_columns:
          - MediaTypeId
    foreign_keys: {}
  - name: Chinook_Playlist
    description: ""
    arguments: {}
    type: Chinook_Playlist
    uniqueness_constraints:
      PlaylistId:
        unique_columns:
          - PlaylistId
    foreign_keys: {}
  - name: Chinook_PlaylistTrack
    description: ""
    arguments: {}
    type: Chinook_PlaylistTrack
    uniqueness_constraints:
      "PlaylistId, TrackId":
        unique_columns:
          - PlaylistId
          - TrackId
    foreign_keys: {}
  - name: Chinook_Track
    description: ""
    arguments: {}
    type: Chinook_Track
    uniqueness_constraints:
      TrackId:
        unique_columns:
          - TrackId
    foreign_keys: {}
functions: []
procedures: []
